File: 1-50/problem14/longest_chain.rb
def chain_length(n)
  return 1 if n == 1
  return 1 + chain_length(n/2) if n % 2 == 0
  return 1 + chain_length(3*n + 1)
end

max_length = 0
max_start = 1
1.upto(1_000_000) do |i|
  puts i
  l = chain_length(i)
  if max_length < l
    max_length = l
    max_start = i
  end
end

puts max_start
puts chain_length(13)

File: 1-50/problem14/longest_chain.c
#include <stdio.h>

int main(int argc, char** argv) {
	unsigned long long i, l, j, max_length = 4, max_start = 1;
	for (i = 2; i < 1000000; i++) {
		j = i;
		l = 1;
		while (j != 1) {
			j = (j%2 == 0) ? j/2 : (j*3 + 1);
			l++;
		}
		if (max_length < l) {
			max_length = l;
			max_start = i;
		}
	}
	
	printf("%lld\n", max_start);
	return 0;
}